src/server/scene/surface_stack.cpp

```diff
@@ -425,6 +425,105 @@ void ms::SurfaceStack::raise(SurfaceSet const& ss)
+            for (auto it = layer.begin(); it != layer.end(); it++)
+            {
+                if (first.count(*it))
+                {
+                    if (first_index < 0)
+                        first_index = it - layer.begin();
+                    first_move.push_back(*it);
+                }
+                else if (second.count(*it))
+                {
+                    if (second_index < 0)
+                        second_index = it - layer.begin();
+                    second_move.push_back(*it);
+                }
+            }
+
+            // Break early if there is nothing to swap here
+            if (second_index < 0 || first_index < 0)
+                continue;
+
+            // Delete the swapped elements  from the list
+            for (auto it = layer.begin(); it != layer.end();)
+            {
+                auto found_first = std::find(first_move.begin(), first_move.end(), *it);
+                auto found_second = std::find(second_move.begin(), second_move.end(), *it);
+                if (found_first != first_move.end() || found_second != second_move.end())
+                     layer.erase(it);
+                else
+                    it++;
+            }
+
+            // Put them back in at the swapped positions
+            if (first_index < second_index)
+            {
+                layer.insert(layer.begin() + first_index, second_move.begin(), second_move.end());
+                layer.insert(layer.begin() + second_index, first_move.begin(), first_move.end());
+            }
+            else
+            {
+                layer.insert(layer.begin() + second_index, first_move.begin(), first_move.end());
+                layer.insert(layer.begin() + first_index, second_move.begin(), second_move.end());
+            }
+        }
+    }
+
+    observers.surfaces_reordered(first);
+}
+
+void ms::SurfaceStack::send_to_back(const mir::scene::SurfaceSet &ss)
+{
+    bool surfaces_reordered{false};
+    {
+        RecursiveWriteLock ul(guard);
+        for (auto& layer : surface_layers)
+        {
+            auto const old_layer = layer;
+
+            // Put all the surfaces to send to the back at the end of the list (preserving order)
+            auto split = std::stable_partition(
+                begin(layer), end(layer),
+                [&](std::weak_ptr<Surface> const& s) { return !ss.count(s); });
+
+            // Make a new vector with only the surfaces to send to the back
+            auto to_send_to_back = std::vector<std::shared_ptr<Surface>>{split, layer.end()};
+
+            // Chop off the surfaces we are moving from the old vector (they are now only in to_send_to_back)
+            layer.erase(split, layer.end());
+
+            // One by one insert to_send_to_back surfaces at the front of the vector
+            for (auto const& surface : to_send_to_back)
+            {
+                layer.insert(layer.begin(), surface);
+            }
```

Comment:
This just looks weird!

1. use std::stable_partition() to move "stuff" to the end
2. copy "stuff"
3. erase "stuff" from the end
4. insert "stuff" at the beginning

Isn't that simply "use std::stable_partition() to move "stuff" to the beginning"?

Reply:
Wow I got to in the weeds here! Much easier :+1: